numpyro/infer/mcmc.py

```diff
@@ -615,6 +621,302 @@ def __init__(self,
         self._algo = 'NUTS'
 
 
+def _get_proposal_loc_and_scale(samples, loc, scale, new_sample):
+    # get loc/scale of q_{-n} (Algorithm 1, line 5 of ref [1]) for n from 1 -> N
+    # these loc/scale will be stacked to the first dim; so
+    #   proposal_loc.shape[0] = proposal_loc.shape[0] = N
+    # Here, we use the numerical stability procedure in Appendix 6 of [1].
+    weight = 1 / samples.shape[0]
```

Comment:
I think this file is getting a bit big, and it may be worth having a separate mcmc module with different files for HMC/NUTS and SA.